### PR TITLE
Update roles header based on CF admin status

### LINKED ISSUE #4: Give platform administrators wider visibility
In order for platform administrators to manage settings and see all content, the proxy should note their group membership in UAA and set the proper roles. 

## Acceptance Criteria
* [ ] GIVEN a user is a member of groups in UAA \
  WHEN this proxy handles a request to Kibana \
  THEN their role is passed on to Opensearch as a header

--- 

## Security considerations

[note any potential changes to security boundaries, practices, documentation, risk that arise directly from this story]

## Implementation sketch

We should do this at the same time as we check the spaces a user has a role in. 

We probably pass this to Opensearch via `headers['x-proxy-roles']` [here](https://github.com/cloud-gov/kibana-cf-auth-proxy/blob/main/kibana_cf_auth_proxy/app.py#L162)

* [x] Add unit test for UAA group membership in `headers['x-proxy-roles']`
* [x] Add code to set UAA group membership as `headers['x-proxy-roles']`

### PR DESCRIPTION
**Pull request by markdboyd**
Closes #4 

## Changes proposed in this pull request:

- Add logic to get a client credentials token from UAA
- Add logic to fetch all of the user's groups from UAA
- Add logic to set `header['x-proxy-roles'] = "admin"` if user is a CF admin
- Remove UAA_TOKEN_URL env var in favor of UAA_BASE_URL 
- Add/update unit tests

## Security considerations

There may be security considerations depending on the soundness of the implementation. For example, are we leaving any holes in our authentication based on the changes in the PR? I think the answer is "no", but a second opinion is necessary.

### CODE REVIEW

File: kibana_cf_auth_proxy/uaa.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        return "Unexpected error", 500
+
+    return response.json()["access_token"]
+
+
+def is_user_cf_admin(user_id, token):
+    with requests.Session() as s:
+        s.headers["Authorization"] = f"Bearer {token}"
+        url = f"{config.UAA_BASE_URL}Users?attributes=groups&filter=id eq '{user_id}'"
+        # there should be only one resource when filtering by ID, so no need for paginating results?
```
Comment (markdboyd) [inline kibana_cf_auth_proxy/uaa.py:32]: I believe this is the case and it was true in my testing, but I'm not sure if there is any edge case for this API that we need to handle

File: tests/unit/test_authn.py [hunk 2]
```diff
@@ -33,7 +24,28 @@ def make_random_token():
...
+def is_valid_auth_code_token_request(request):
+    data = request.text
+    data = parse.parse_qs(data)
+    assert data["grant_type"][0] == "authorization_code"
+    assert data["code"][0] == "1234"
+    assert data["redirect_uri"][0] == "http://localhost/cb"
+
+
+def is_client_credentials_token_request(request):
+    return "grant_type=client_credentials" in request.text
```
Comment (markdboyd) [inline tests/unit/test_authn.py:40]: Refactoring to separate the mocking matchers and the test assertions was necessary because now we have two requests to the mocked token endpoint during the "happy path" test:

- One to get the access token as the user when for the `authorization_code` grant type
- One to get the access token as the client when for the `client_credentials` grant type

requests_mock works by running all possible matchers defined for the same endpoint in order to determine how to handle that request. So if I left the `assert` statements in the matcher functions, the test would immediately fail when making the client credentials token request, because those assertions are only valid for the authorization code token request.

File: kibana_cf_auth_proxy/config.py [hunk 3]
```diff
@@ -15,6 +15,7 @@ class Config:
     def __init__(self):
         self.env_parser = Env()
         self.PORT = self.env_parser.int("PORT", 8080)
+        self.CF_ADMIN_GROUP_NAME = "cloud_controller.admin"
```
[the remark this answers is not in the document]
Comment (bengerman13) [reply kibana_cf_auth_proxy/config.py:18]: I think we should make this a variable since the intent is to have this be reusable by other CF installations who may have a different opinion about what makes someone a superuser

File: kibana_cf_auth_proxy/uaa.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    except:
+        return "Unexpected error", 500
+
+    return response.json()["access_token"]
+
+
+def is_user_cf_admin(user_id, token):
+    with requests.Session() as s:
+        s.headers["Authorization"] = f"Bearer {token}"
+        url = f"{config.UAA_BASE_URL}Users?attributes=groups&filter=id eq '{user_id}'"
```
Comment (bengerman13) [inline kibana_cf_auth_proxy/uaa.py:31]: I think this would be a little safer and read a little better if we made a dict of the query parameters and made `requests` handle parsing them onto the the querystring.

Comment (bengerman13) [inline kibana_cf_auth_proxy/uaa.py:31]: Is there a reason to prefer 
`GET /Users?filter=id eq $id`
over
`GET /Users/$id`

**Comment by bengerman13** [comment]
I added a couple nits and thoughts, otherwise it looks really good!

**Review by bengerman13 — APPROVED** [review]
